soc/xtensa/intel_adsp/ace_v1x/power.c

```diff
@@ -26,29 +26,194 @@
  * (each bit corresponds to one ebb)
  * @param response_to_ipc       flag if ipc response should be send during power down
  */
-extern void ace_power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask,
+extern void power_down(bool disable_lpsram, uint32_t *hpsram_pg_mask,
 			   bool response_to_ipc);
 
+static ALWAYS_INLINE uint32_t read_intenable(void)
+{
+	uint32_t intenable;
+
+	__asm__ volatile("rsr.intenable %0" : "=r"(intenable));
+	return intenable;
+}
+
+static ALWAYS_INLINE uint32_t read_threadptr(void)
+{
+	uint32_t threadptr;
+
+	__asm__ volatile(
+		"rur.threadptr %0" : "=a"(threadptr));
+	return threadptr;
+}
+
+static ALWAYS_INLINE void write_threadptr(uint32_t threadptr)
+{
+	__asm__ volatile(
+		"wur.threadptr %0" : : "a" (threadptr) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_vecbase(void)
+{
+	uint32_t vecbase;
+
+	__asm__ volatile("rsr.vecbase %0"
+		: "=a"(vecbase));
+	return vecbase;
+}
+
+static ALWAYS_INLINE void write_vecbase(uint32_t vecbase)
+{
+	__asm__ volatile("wsr.vecbase %0"
+		: : "a"(vecbase) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave2(void)
+{
+	uint32_t excsave2;
+
+	__asm__ __volatile__("rsr.excsave2 %0"
+		: "=a"(excsave2));
+	return excsave2;
+}
+
+static ALWAYS_INLINE void write_excsave2(uint32_t excsave2)
+{
+	__asm__ __volatile__("wsr.excsave2 %0"
+		: : "a"(excsave2) : "memory");
+}
+
+static ALWAYS_INLINE uint32_t read_excsave3(void)
+{
+	uint32_t excsave3;
+
+	__asm__ __volatile__("rsr.excsave3 %0"
+		: "=a"(excsave3));
+	return excsave3;
+}
+
+static ALWAYS_INLINE void write_excsave3(uint32_t excsave3)
+{
+	__asm__ __volatile__("wsr.excsave3 %0"
+		: : "a"(excsave3) : "memory");
+}
```

Comment:
@ceolin I have doubts about the location of this code, but I haven't found a better place. Any suggestions?

Reply:
@ceolin alternatively, I can use macros **RSR** and **WSR** (from _arch/xtensa/arch_inlines.h_). I would only have to add support for user registers (_RUR.THREADPTR_).

Reply:
The place is fine, these functions that don't look good. Can't you group them by functionality instead of per instruction ?
For example, you have `write_threadptr`, `write_vecbase`, `write_excsave2` and `write_excsave3` and they are used only once and together:
```
+		write_threadptr(core_desc[core_id].thread_ptr);
+		write_vecbase(core_desc[core_id].vecbase);
+		write_excsave2(core_desc[core_id].excsave2);
+		write_excsave3(core_desc[core_id].excsave3);
```

Couldn't they be just one function like `*save_context*()` ?

Same apply for other similar functions

Reply:
I replaced this with usage of RSR/WSR macros. I think this looks better, less code added :)